// @iapps/ngx-dhis2-menu/esm5/lib/containers/menu/menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { MenuService } from '../../services/menu.service';
import { SystemStateService } from '../../services/system-state.service';
import { MENU_BACKGROUND_COLORS } from '../../constants/menu-background-colors';
var MenuComponent = /** @class */ (function () {
    function MenuComponent(menuService, systemStatusService) {
        this.menuService = menuService;
        this.systemStatusService = systemStatusService;
        this.rootUrl = '../../../';
        this.menuLoading = true;
        this.menuLoadingFail = false;
        this.loggedIn$ = this.systemStatusService.getLoginStatus();
        this.online = false;
        this.menuNotification = '';
        this.wasOffline = true;
        this.backgroundColor = '#ECECEC';
        this.showSidebar = false;
    }
    /**
     * @return {?}
     */
    MenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.systemStatusService.checkOnlineStatus().subscribe(function (onlineStatus) {
            _this.online = onlineStatus;
            if (_this.online && _this.wasOffline) {
                _this.menuNotification = 'You are online';
                _this.wasOffline = false;
                /**
                         * Hide notification status after sometimes
                         */
                setTimeout(function () {
                    _this.menuNotification = '';
                }, 3000);
                /**
                         * Load system settings if failed
                         */
                if (_this.menuLoadingFail) {
                    _this.menuLoading = true;
                    _this.getSystemSettings();
                }
            }
            else if (!_this.online) {
                _this.menuNotification = 'You are offline';
                _this.wasOffline = true;
            }
        });
        this.getSystemSettings();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    MenuComponent.prototype.toggleSideBar = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        this.showSidebar = !this.showSidebar;
    };
    /**
     * @return {?}
     */
    MenuComponent.prototype.getSystemSettings = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.menuService.getSystemSettings(this.rootUrl).subscribe(function (settings) {
            if (settings) {
                _this.applicationTitle = settings['applicationTitle'];
                // get system current background style
                var /** @type {?} */ colorName = settings.hasOwnProperty('currentStyle')
                    ? settings['currentStyle'].split('/')[0]
                    : settings.hasOwnProperty('keyStyle')
                        ? settings['keyStyle'].split('/')[0]
                        : 'blue';
                _this.backgroundColor =
                    MENU_BACKGROUND_COLORS[colorName];
                _this.contextPath = settings.contextPath ? settings.contextPath + '/' : '';
            }
            _this.menuLoading = false;
            _this.menuLoadingFail = false;
        }, function () {
            _this.menuLoading = false;
            _this.menuLoadingFail = true;
        });
    };
    MenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-dhis2-menu',
                    template: "<div class=\"menu-container\">\n\n  <!-- notification area -->\n  <div class=\"menu-notification-area\" *ngIf=\"menuNotification !== ''\" [@open]>\n    {{ menuNotification }}\n  </div>\n  <!-- end of notification area -->\n  <!-- login form area -->\n  <div class=\"login-form-container\" *ngIf=\"!(loggedIn$ | async)\">\n    <app-login-form></app-login-form>\n  </div>\n  <!-- end of login form area -->\n\n  <!--menu loading area-->\n  <div class=\"menu-loader\" *ngIf=\"menuLoading\">\n    <app-menu-loader></app-menu-loader>\n  </div>\n  <!--end of menu loading area-->\n\n  <!--menu loading failed area-->\n  <div class=\"menu-content\" *ngIf=\"!menuLoading && menuLoadingFail\" style=\"background-color: #eeeeee\">\n    &nbsp;\n  </div>\n  <!--end of menu loading failed area-->\n\n  <!--menu content area-->\n  <div class=\"menu-content\" [style.background-color]=\"backgroundColor\" *ngIf=\"!menuLoading && !menuLoadingFail\">\n    <!--menu left block-->\n    <div class=\"menu-left-block\">\n\n      <!--menu brand area-->\n      <div class=\"menu-brand\">\n        <a [href]=\"rootUrl + 'dhis-web-commons-about/redirect.action'\">\n          <img [src]=\"rootUrl + 'api/staticContent/logo_banner'\" alt=\"\">\n        </a>\n      </div>\n      <!--end of menu brand area-->\n\n      <!--menu title area-->\n      <div class=\"menu-title-block\">\n        <a [href]=\"rootUrl + 'dhis-web-commons-about/redirect.action'\" class=\"menu-title\">\n          {{applicationTitle}}\n        </a>\n      </div>\n      <!--end of menu title area-->\n    </div>\n    <!--end of menu left block-->\n\n    <!--menu right block-->\n    <div class=\"menu-right-block\">\n      <div class=\"menu-notification-block\">\n        <app-menu-notification [rootUrl]=\"rootUrl\"></app-menu-notification>\n      </div>\n      <!--menu search block-->\n      <div class=\"menu-search-block\">\n        <app-menu-search></app-menu-search>\n      </div>\n      <!--end of menu search block-->\n\n      <!--menu profile block-->\n      <div class=\"menu-profile-block\">\n        <app-menu-profile [rootUrl]=\"rootUrl\" [contextPath]=\"contextPath\" [backgroundColor]=\"backgroundColor\"></app-menu-profile>\n      </div>\n      <!--end of menu profile block-->\n\n      <!--menu sidebar toggle button-->\n      <div\n        class=\"menu-toggle-button\"\n      >\n        <a\n          (click)=\"toggleSideBar($event)\"\n        >\n          <svg\n            class=\"menu-toggle-icon\"\n            viewBox=\"0 0 24 24\"\n          >\n            <path d=\"M4 8h4V4H4v4zm6 12h4v-4h-4v4zm-6 0h4v-4H4v4zm0-6h4v-4H4v4zm6 0h4v-4h-4v4zm6-10v4h4V4h-4zm-6 4h4V4h-4v4zm6 6h4v-4h-4v4zm0 6h4v-4h-4v4z\"></path>\n          </svg>\n        </a>\n      </div>\n      <!--end of menu sidebar toggle button-->\n\n      <!--sidebar menu block-->\n      <div class=\"sidebar-menu-nav\" *ngIf=\"showSidebar\">\n        <app-menu-side-bar></app-menu-side-bar>\n      </div>\n      <!--end of sidebar menu block-->\n      <div></div>\n    </div>\n    <!--end of menu right block-->\n  </div>\n  <!--end of menu content area-->\n</div>\n",
                    styles: [".menu-container{position:fixed;left:0;top:0;right:0;z-index:1400;height:48px}.menu-loader{height:100%}.menu-content{display:flex;width:100%;height:100%;justify-content:space-between}.menu-left-block,.menu-right-block{display:flex}.menu-brand{width:155px;height:100%;display:flex;justify-content:center;align-items:center}.menu-brand>a>img{display:block;margin:auto;max-height:44px;max-width:155px}.menu-title-block{overflow:hidden;white-space:nowrap;color:#fff;align-items:center;justify-items:center;display:flex;padding-right:1rem}.menu-title{font-size:1rem;font-weight:700;color:#fff;text-decoration:none;text-overflow:ellipsis;min-width:1px;overflow:hidden;white-space:nowrap}.menu-search-block{height:100%}.login-form-container{position:fixed;z-index:1500;top:2px;left:20%;min-height:42px;border-color:#f0c36d;background-color:#f9edbe;border-radius:3px}.menu-notification-area{position:fixed;z-index:1500;top:7px;left:20%;height:auto;border-color:#f0c36d;background-color:#f9edbe;border-radius:3px;padding:5px;font-size:15px}.menu-toggle-button{display:none;flex-direction:row;justify-content:flex-end;padding-right:10px;padding-top:6px}.menu-toggle-icon{display:inline-block;height:35px;width:35px;fill:#fff}.sidebar-menu-nav{position:absolute;top:100%;width:400px;height:100vh;right:0;z-index:900;font-size:14px;text-align:left;list-style:none;background-color:#fff;-webkit-background-clip:padding-box;transition:width .5s cubic-bezier(.39,-.04,.2,1);background-clip:padding-box;box-shadow:0 6px 12px rgba(0,0,0,.175)}@media (max-width:400px){.sidebar-menu-nav{width:95%}}@media (max-width:500px){.login-form-container{top:50px;left:2%;width:96%;height:auto}.menu-notification-area{left:2%;right:2%;width:96%;height:auto}}@media (max-width:999px){.menu-profile-block,.menu-search-block,.menu-title-block{display:none}.menu-brand{justify-content:flex-start;margin-left:15px}.menu-toggle-button{display:flex}}.menu-notification-block{display:flex;align-items:center;height:100%;margin-right:30px}@media (min-width:1010px){.sidebar-menu-nav{display:none!important}}"],
                    animations: [
                        trigger('open', [
                            state('in', style({
                                opacity: 1
                            })),
                            transition('void => *', [
                                style({
                                    opacity: 0
                                }),
                                animate(700)
                            ]),
                            transition('* => void', [
                                animate(400),
                                style({
                                    opacity: 0
                                })
                            ])
                        ])
                    ]
                },] },
    ];
    /** @nocollapse */
    MenuComponent.ctorParameters = function () { return [
        { type: MenuService },
        { type: SystemStateService }
    ]; };
    return MenuComponent;
}());
export { MenuComponent };
function MenuComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    MenuComponent.prototype.rootUrl;
    /** @type {?} */
    MenuComponent.prototype.applicationTitle;
    /** @type {?} */
    MenuComponent.prototype.backgroundColor;
    /** @type {?} */
    MenuComponent.prototype.menuLoading;
    /** @type {?} */
    MenuComponent.prototype.menuLoadingFail;
    /** @type {?} */
    MenuComponent.prototype.loggedIn$;
    /** @type {?} */
    MenuComponent.prototype.online;
    /** @type {?} */
    MenuComponent.prototype.menuNotification;
    /** @type {?} */
    MenuComponent.prototype.wasOffline;
    /** @type {?} */
    MenuComponent.prototype.showSidebar;
    /** @type {?} */
    MenuComponent.prototype.contextPath;
    /** @type {?} */
    MenuComponent.prototype.menuService;
    /** @type {?} */
    MenuComponent.prototype.systemStatusService;
}
