// @iapps/ngx-dhis2-http-client/esm2015/lib/services/index-db.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import * as _ from 'lodash';
import { switchMap, mergeMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class IndexDbService {
    constructor() {
        this._indexedDB = indexedDB;
        this._dbName = 'db'; // by default
    }
    /**
     * @param {?} dbName
     * @return {?}
     */
    setName(dbName) {
        if (dbName) {
            this._dbName = dbName;
        }
    }
    /**
     * Function to update existing data in the db
     * @param {?} schema
     * @param {?} data
     * @return {?}
     */
    put(schema, data) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(switchMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                if (_.isArray(data)) {
                    _.each(data, dataItem => {
                        store.put(dataItem);
                    });
                }
                else {
                    store.put(data);
                }
                transaction.oncomplete = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
                db.onerror = (errorEvent) => {
                    db.close();
                    observer.error(errorEvent.target.errorCode);
                };
            });
        });
    }
    /**
     * Function to add new data in the db
     * @param {?} schema
     * @param {?} data
     * @return {?}
     */
    post(schema, data) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(mergeMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                // Check if data is array or an object
                if (_.isArray(data)) {
                    _.each(data, dataItem => {
                        store.add(dataItem);
                    });
                }
                else {
                    store.add(data);
                }
                transaction.oncomplete = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
                db.onerror = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
            });
        });
    }
    /**
     * Function to get value for a particular key from the table
     * @param {?} schema
     * @param {?} key
     * @return {?}
     */
    findOne(schema, key) {
        return new Observable((observer) => {
            this.create(schema)
                .pipe(mergeMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                const /** @type {?} */ request = store.get(key);
                request.onsuccess = successEvent => {
                    observer.next(successEvent.target.result);
                    db.close();
                    observer.complete();
                };
                db.onerror = (event) => {
                    console.log(event.target.errorCode);
                    db.close();
                    observer.error(event.target.errorCode);
                };
            });
        });
    }
    /**
     * Function to find all values for a particular table in the db
     * @param {?} schema
     * @return {?}
     */
    findAll(schema) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(switchMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                const /** @type {?} */ storeIndex = store.index(schema.keyPath);
                const /** @type {?} */ request = storeIndex.openCursor();
                let /** @type {?} */ results = [];
                request.onsuccess = successEvent => {
                    const /** @type {?} */ cursor = successEvent.target.result;
                    if (cursor) {
                        results = [...results, cursor.value];
                        cursor.continue();
                    }
                    else {
                        observer.next(results);
                        db.close();
                        observer.complete();
                    }
                };
                db.onerror = (event) => {
                    db.close();
                    observer.error(event.target.errorCode);
                };
            });
        });
    }
    /**
     * Delete data from the db
     * @param {?} schema
     * @param {?} key
     * @return {?}
     */
    delete(schema, key) {
        return Observable.create((observer) => {
            this.open().subscribe((db) => {
                if (!db.objectStoreNames.contains(schema.name)) {
                    const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                    const /** @type {?} */ store = transaction.objectStore(schema.name);
                    if (_.isArray(key)) {
                        _.each(key, keyItem => {
                            store.delete(keyItem);
                        });
                    }
                    else {
                        store.delete(key);
                    }
                    transaction.oncomplete = () => {
                        observer.next(key);
                        db.close();
                        observer.complete();
                    };
                    db.onerror = (errorEvent) => {
                        db.close();
                        observer.error(errorEvent.target.errorCode);
                    };
                }
            });
        });
    }
    /**
     * Create a schema/table
     * @param {?} schema
     * @return {?}
     */
    create(schema) {
        return Observable.create((observer) => {
            this.open().subscribe((requestResult) => {
                if (!requestResult.objectStoreNames.contains(schema.name)) {
                    /**
                               * Close database to allow opening with new version
                               */
                    requestResult.close();
                    /**
                     * Set a new connection with new version to allows creating schema
                     */
                    const /** @type {?} */ request = this._indexedDB.open(this._dbName, requestResult.version + 1);
                    request.onupgradeneeded = event => {
                        const /** @type {?} */ db = event.target.result;
                        if (!db.objectStoreNames.contains(schema.name)) {
                            /**
                             * Create corresponding schema
                             */
                            const /** @type {?} */ store = db.createObjectStore(schema.name, {
                                keyPath: schema.keyPath
                            });
                            store.createIndex(schema.keyPath, schema.keyPath, {
                                unique: true
                            });
                            /**
                                           * Create indexes for schema if supplied
                                           */
                            if (schema.indexes) {
                                _.each(schema.indexes, (schemaIndex) => {
                                    store.createIndex(schemaIndex, schemaIndex);
                                });
                            }
                            /**
                                           * Add data if supplied
                                           */
                            if (schema.data) {
                                _.each(schema.data, (dataItem) => {
                                    store.put(dataItem);
                                });
                            }
                        }
                        observer.next('done');
                        observer.complete();
                    };
                    request.onerror = errorEvent => {
                        observer.error(errorEvent.target.errorCode);
                    };
                    request.onsuccess = successEvent => {
                        successEvent.target.result.close();
                        observer.next('done');
                        observer.complete();
                    };
                }
                else {
                    observer.next('done');
                    observer.complete();
                }
            });
        });
    }
    /**
     * Delete the whole database
     * @return {?}
     */
    clear() {
        return Observable.create((observer) => {
            const /** @type {?} */ request = this._indexedDB.deleteDatabase(this._dbName);
            request.onsuccess = () => {
                observer.next('done');
                observer.complete();
            };
            request.onerror = errorEvent => {
                observer.error(errorEvent.target.errorCode);
            };
            request.onblocked = blockEvent => {
                observer.error(blockEvent.target.errorCode);
            };
        });
    }
    /**
     * @return {?}
     */
    open() {
        return Observable.create((observer) => {
            const /** @type {?} */ request = this._indexedDB.open(this._dbName);
            request.onsuccess = successEvent => {
                observer.next(successEvent.target.result);
                observer.complete();
            };
            request.onerror = errorEvent => observer.error(errorEvent.target.errorCode);
        });
    }
}
IndexDbService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
IndexDbService.ctorParameters = () => [];
/** @nocollapse */ IndexDbService.ngInjectableDef = i0.defineInjectable({ factory: function IndexDbService_Factory() { return new IndexDbService(); }, token: IndexDbService, providedIn: "root" });
function IndexDbService_tsickle_Closure_declarations() {
    /** @type {?} */
    IndexDbService.prototype._indexedDB;
    /** @type {?} */
    IndexDbService.prototype._dbName;
}
