// @iapps/ngx-dhis2-http-client/esm2015/lib/services/system-info.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { forkJoin, of } from 'rxjs';
import { ManifestService } from './manifest.service';
import { map, switchMap, tap } from 'rxjs/internal/operators';
import { HttpClient } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "./manifest.service";
import * as i2 from "@angular/common/http";
export class SystemInfoService {
    /**
     * @param {?} manifestService
     * @param {?} httpClient
     */
    constructor(manifestService, httpClient) {
        this.manifestService = manifestService;
        this.httpClient = httpClient;
        this._systemInfoLoaded = false;
    }
    /**
     * @return {?}
     */
    getSystemInfo() {
        return this._systemInfoLoaded ? of(this._systemInfo) :
            this.manifestService.getRootUrl().pipe(switchMap((rootUrl) => forkJoin(this.httpClient.get(`${rootUrl}api/system/info`), this.httpClient.get(`${rootUrl}api/systemSettings`)).pipe(map((res) => {
                return Object.assign({}, res[0], res[1]);
            }), tap((systemInfo) => {
                this._systemInfo = systemInfo;
                this._systemInfoLoaded = true;
            }))));
    }
    /**
     * @return {?}
     */
    getSystemVersion() {
        return this.getSystemInfo().pipe(map((systemInfo) => {
            if (!systemInfo) {
                return 0;
            }
            const /** @type {?} */ splitedVersion = systemInfo.version ? systemInfo.version.split('.') : [];
            return parseInt(splitedVersion[1], 10) || 0;
        }));
    }
}
SystemInfoService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
SystemInfoService.ctorParameters = () => [
    { type: ManifestService },
    { type: HttpClient }
];
/** @nocollapse */ SystemInfoService.ngInjectableDef = i0.defineInjectable({ factory: function SystemInfoService_Factory() { return new SystemInfoService(i0.inject(i1.ManifestService), i0.inject(i2.HttpClient)); }, token: SystemInfoService, providedIn: "root" });
function SystemInfoService_tsickle_Closure_declarations() {
    /** @type {?} */
    SystemInfoService.prototype._systemInfoLoaded;
    /** @type {?} */
    SystemInfoService.prototype._systemInfo;
    /** @type {?} */
    SystemInfoService.prototype.manifestService;
    /** @type {?} */
    SystemInfoService.prototype.httpClient;
}
