// @iapps/ngx-dhis2-http-client/fesm2015/iapps-ngx-dhis2-http-client.js
import { Injectable, defineInjectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of, forkJoin, Observable, throwError } from 'rxjs';
import { tap, catchError, map, switchMap, mergeMap as mergeMap$1 } from 'rxjs/internal/operators';
import { isArray, each } from 'lodash';
import { switchMap as switchMap$1, mergeMap } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class ManifestService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this._defaultRootUrl = '../../../';
        this._manifestLoaded = false;
    }
    /**
     * @return {?}
     */
    getManifest() {
        return this._manifestLoaded ? of(this._manifest) : this.httpClient.get('manifest.webapp').pipe(catchError(() => {
            console.warn('Manifest file could not be loaded, default options have been used instead');
            return of(null);
        }), tap((manifest) => {
            this._manifest = manifest;
            this._manifestLoaded = true;
        }));
    }
    /**
     * @return {?}
     */
    getRootUrl() {
        return this.getManifest().pipe(map((manifest) => {
            if (!manifest) {
                return this._defaultRootUrl;
            }
            return manifest.activities && manifest.activities.dhis && manifest.activities.dhis.href ?
                manifest.activities.dhis.href : this._defaultRootUrl;
        }));
    }
}
ManifestService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
ManifestService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ ManifestService.ngInjectableDef = defineInjectable({ factory: function ManifestService_Factory() { return new ManifestService(inject(HttpClient)); }, token: ManifestService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class SystemInfoService {
    /**
     * @param {?} manifestService
     * @param {?} httpClient
     */
    constructor(manifestService, httpClient) {
        this.manifestService = manifestService;
        this.httpClient = httpClient;
        this._systemInfoLoaded = false;
    }
    /**
     * @return {?}
     */
    getSystemInfo() {
        return this._systemInfoLoaded ? of(this._systemInfo) :
            this.manifestService.getRootUrl().pipe(switchMap((rootUrl) => forkJoin(this.httpClient.get(`${rootUrl}api/system/info`), this.httpClient.get(`${rootUrl}api/systemSettings`)).pipe(map((res) => {
                return Object.assign({}, res[0], res[1]);
            }), tap((systemInfo) => {
                this._systemInfo = systemInfo;
                this._systemInfoLoaded = true;
            }))));
    }
    /**
     * @return {?}
     */
    getSystemVersion() {
        return this.getSystemInfo().pipe(map((systemInfo) => {
            if (!systemInfo) {
                return 0;
            }
            const /** @type {?} */ splitedVersion = systemInfo.version ? systemInfo.version.split('.') : [];
            return parseInt(splitedVersion[1], 10) || 0;
        }));
    }
}
SystemInfoService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
SystemInfoService.ctorParameters = () => [
    { type: ManifestService },
    { type: HttpClient }
];
/** @nocollapse */ SystemInfoService.ngInjectableDef = defineInjectable({ factory: function SystemInfoService_Factory() { return new SystemInfoService(inject(ManifestService), inject(HttpClient)); }, token: SystemInfoService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ HTTP_CONFIG = {
    includeVersionNumber: false,
    preferPreviousApiVersion: false,
    useRootUrl: false,
    useIndexDb: false,
    isExternalLink: false
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class IndexDbService {
    constructor() {
        this._indexedDB = indexedDB;
        this._dbName = 'db'; // by default
    }
    /**
     * @param {?} dbName
     * @return {?}
     */
    setName(dbName) {
        if (dbName) {
            this._dbName = dbName;
        }
    }
    /**
     * Function to update existing data in the db
     * @param {?} schema
     * @param {?} data
     * @return {?}
     */
    put(schema, data) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(switchMap$1(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                if (isArray(data)) {
                    each(data, dataItem => {
                        store.put(dataItem);
                    });
                }
                else {
                    store.put(data);
                }
                transaction.oncomplete = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
                db.onerror = (errorEvent) => {
                    db.close();
                    observer.error(errorEvent.target.errorCode);
                };
            });
        });
    }
    /**
     * Function to add new data in the db
     * @param {?} schema
     * @param {?} data
     * @return {?}
     */
    post(schema, data) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(mergeMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                // Check if data is array or an object
                if (isArray(data)) {
                    each(data, dataItem => {
                        store.add(dataItem);
                    });
                }
                else {
                    store.add(data);
                }
                transaction.oncomplete = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
                db.onerror = () => {
                    observer.next(data);
                    db.close();
                    observer.complete();
                };
            });
        });
    }
    /**
     * Function to get value for a particular key from the table
     * @param {?} schema
     * @param {?} key
     * @return {?}
     */
    findOne(schema, key) {
        return new Observable((observer) => {
            this.create(schema)
                .pipe(mergeMap(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                const /** @type {?} */ request = store.get(key);
                request.onsuccess = successEvent => {
                    observer.next(successEvent.target.result);
                    db.close();
                    observer.complete();
                };
                db.onerror = (event) => {
                    console.log(event.target.errorCode);
                    db.close();
                    observer.error(event.target.errorCode);
                };
            });
        });
    }
    /**
     * Function to find all values for a particular table in the db
     * @param {?} schema
     * @return {?}
     */
    findAll(schema) {
        return Observable.create((observer) => {
            this.create(schema)
                .pipe(switchMap$1(() => this.open()))
                .subscribe((db) => {
                const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                const /** @type {?} */ store = transaction.objectStore(schema.name);
                const /** @type {?} */ storeIndex = store.index(schema.keyPath);
                const /** @type {?} */ request = storeIndex.openCursor();
                let /** @type {?} */ results = [];
                request.onsuccess = successEvent => {
                    const /** @type {?} */ cursor = successEvent.target.result;
                    if (cursor) {
                        results = [...results, cursor.value];
                        cursor.continue();
                    }
                    else {
                        observer.next(results);
                        db.close();
                        observer.complete();
                    }
                };
                db.onerror = (event) => {
                    db.close();
                    observer.error(event.target.errorCode);
                };
            });
        });
    }
    /**
     * Delete data from the db
     * @param {?} schema
     * @param {?} key
     * @return {?}
     */
    delete(schema, key) {
        return Observable.create((observer) => {
            this.open().subscribe((db) => {
                if (!db.objectStoreNames.contains(schema.name)) {
                    const /** @type {?} */ transaction = db.transaction(schema.name, 'readwrite');
                    const /** @type {?} */ store = transaction.objectStore(schema.name);
                    if (isArray(key)) {
                        each(key, keyItem => {
                            store.delete(keyItem);
                        });
                    }
                    else {
                        store.delete(key);
                    }
                    transaction.oncomplete = () => {
                        observer.next(key);
                        db.close();
                        observer.complete();
                    };
                    db.onerror = (errorEvent) => {
                        db.close();
                        observer.error(errorEvent.target.errorCode);
                    };
                }
            });
        });
    }
    /**
     * Create a schema/table
     * @param {?} schema
     * @return {?}
     */
    create(schema) {
        return Observable.create((observer) => {
            this.open().subscribe((requestResult) => {
                if (!requestResult.objectStoreNames.contains(schema.name)) {
                    /**
                               * Close database to allow opening with new version
                               */
                    requestResult.close();
                    /**
                     * Set a new connection with new version to allows creating schema
                     */
                    const /** @type {?} */ request = this._indexedDB.open(this._dbName, requestResult.version + 1);
                    request.onupgradeneeded = event => {
                        const /** @type {?} */ db = event.target.result;
                        if (!db.objectStoreNames.contains(schema.name)) {
                            /**
                             * Create corresponding schema
                             */
                            const /** @type {?} */ store = db.createObjectStore(schema.name, {
                                keyPath: schema.keyPath
                            });
                            store.createIndex(schema.keyPath, schema.keyPath, {
                                unique: true
                            });
                            /**
                                           * Create indexes for schema if supplied
                                           */
                            if (schema.indexes) {
                                each(schema.indexes, (schemaIndex) => {
                                    store.createIndex(schemaIndex, schemaIndex);
                                });
                            }
                            /**
                                           * Add data if supplied
                                           */
                            if (schema.data) {
                                each(schema.data, (dataItem) => {
                                    store.put(dataItem);
                                });
                            }
                        }
                        observer.next('done');
                        observer.complete();
                    };
                    request.onerror = errorEvent => {
                        observer.error(errorEvent.target.errorCode);
                    };
                    request.onsuccess = successEvent => {
                        successEvent.target.result.close();
                        observer.next('done');
                        observer.complete();
                    };
                }
                else {
                    observer.next('done');
                    observer.complete();
                }
            });
        });
    }
    /**
     * Delete the whole database
     * @return {?}
     */
    clear() {
        return Observable.create((observer) => {
            const /** @type {?} */ request = this._indexedDB.deleteDatabase(this._dbName);
            request.onsuccess = () => {
                observer.next('done');
                observer.complete();
            };
            request.onerror = errorEvent => {
                observer.error(errorEvent.target.errorCode);
            };
            request.onblocked = blockEvent => {
                observer.error(blockEvent.target.errorCode);
            };
        });
    }
    /**
     * @return {?}
     */
    open() {
        return Observable.create((observer) => {
            const /** @type {?} */ request = this._indexedDB.open(this._dbName);
            request.onsuccess = successEvent => {
                observer.next(successEvent.target.result);
                observer.complete();
            };
            request.onerror = errorEvent => observer.error(errorEvent.target.errorCode);
        });
    }
}
IndexDbService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
IndexDbService.ctorParameters = () => [];
/** @nocollapse */ IndexDbService.ngInjectableDef = defineInjectable({ factory: function IndexDbService_Factory() { return new IndexDbService(); }, token: IndexDbService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxDhis2HttpClientService {
    /**
     * @param {?} httpClient
     * @param {?} manifestService
     * @param {?} systemInfoService
     * @param {?} indexDbService
     */
    constructor(httpClient, manifestService, systemInfoService, indexDbService) {
        this.httpClient = httpClient;
        this.manifestService = manifestService;
        this.systemInfoService = systemInfoService;
        this.indexDbService = indexDbService;
    }
    /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    get(url, httpConfig) {
        const /** @type {?} */ newHttpConfig = Object.assign({}, HTTP_CONFIG, httpConfig);
        // Make a call directly from url if is external one
        if (newHttpConfig.isExternalLink) {
            return this.httpClient.get(url);
        }
        // load from index db
        if (newHttpConfig.useIndexDb) {
            return this._fetchFromIndexDb(newHttpConfig).pipe(mergeMap$1((indexDbResponse) => indexDbResponse
                ? of(indexDbResponse)
                : this._get(url, newHttpConfig).pipe(mergeMap$1((response) => this._updateIndexDb(newHttpConfig, response)))));
        }
        return this._get(url, newHttpConfig);
    }
    /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    post(url, data, httpConfig) {
        const /** @type {?} */ newHttpConfig = Object.assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap$1(rootUrl => this.httpClient
            .post(rootUrl + url, data)
            .pipe(mergeMap$1((response) => {
            const /** @type {?} */ dataId = response.uid;
            return newHttpConfig.useIndexDb
                ? dataId
                    ? this._updateIndexDb(newHttpConfig, Object.assign({}, data, { id: dataId }), 'CREATE').pipe(map(() => response))
                    : of(response)
                : of(response);
        }))
            .pipe(catchError(this._handleError))), catchError(this._handleError));
    }
    /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    put(url, data, httpConfig) {
        const /** @type {?} */ newHttpConfig = Object.assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap$1(rootUrl => this.httpClient.put(rootUrl + url, data).pipe(mergeMap$1((response) => newHttpConfig.useIndexDb
            ? this._updateIndexDb(newHttpConfig, data, 'UPDATE').pipe(map(() => response))
            : of(response)), catchError(this._handleError))), catchError(this._handleError));
    }
    /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    delete(url, httpConfig) {
        const /** @type {?} */ newHttpConfig = Object.assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap$1(rootUrl => this.httpClient
            .delete(rootUrl + url)
            .pipe(mergeMap$1((response) => newHttpConfig.useIndexDb
            ? this._updateIndexDb(newHttpConfig, null, 'DELETE').pipe(map(() => response))
            : of(response)))
            .pipe(catchError(this._handleError))), catchError(this._handleError));
    }
    /**
     * @param {?} url
     * @param {?} httpConfig
     * @return {?}
     */
    _get(url, httpConfig) {
        return this._getSanitizedRootUrl(httpConfig).pipe(mergeMap$1(rootUrl => this.httpClient.get(rootUrl + url).pipe(catchError(this._handleError))), catchError(this._handleError));
    }
    /**
     * @param {?} err
     * @return {?}
     */
    _handleError(err) {
        let /** @type {?} */ error = null;
        if (err.error instanceof ErrorEvent) {
            // A client-side or network error occurred. Handle it accordingly.
            error = {
                message: err.error,
                status: err.status,
                statusText: err.statusText
            };
        }
        else {
            // The backend returned an unsuccessful response code.
            // The response body may contain clues as to what went wrong,
            error = {
                message: err.error instanceof Object
                    ? err.error.message
                    : err.error || err.message,
                status: err.status,
                statusText: err.statusText
            };
        }
        return throwError(error);
    }
    /**
     * @param {?} httpConfig
     * @return {?}
     */
    _getSanitizedRootUrl(httpConfig) {
        return httpConfig.useRootUrl
            ? this.manifestService.getRootUrl()
            : this._getApiRootUrl(httpConfig.includeVersionNumber, httpConfig.preferPreviousApiVersion);
    }
    /**
     * @param {?=} includeVersionNumber
     * @param {?=} preferPreviousVersion
     * @return {?}
     */
    _getApiRootUrl(includeVersionNumber = false, preferPreviousVersion = false) {
        const /** @type {?} */ rootUrlPromise = this.manifestService.getRootUrl().pipe(switchMap(rootUrl => this.systemInfoService.getSystemVersion().pipe(map((version) => {
            return {
                rootUrl,
                version: version - 1 <= 25 ? version + 1 : version
            };
        }))));
        return rootUrlPromise.pipe(map((urlInfo) => `${urlInfo.rootUrl}api/${includeVersionNumber && !preferPreviousVersion
            ? urlInfo.version + '/'
            : preferPreviousVersion
                ? urlInfo.version
                    ? urlInfo.version - 1 + '/'
                    : ''
                : ''}`));
    }
    /**
     * @param {?} httpConfig
     * @param {?} requestData
     * @param {?=} action
     * @return {?}
     */
    _updateIndexDb(httpConfig, requestData, action = 'CREATE') {
        const /** @type {?} */ indexDbSchema = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.schema
            : null;
        const /** @type {?} */ arrayKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.arrayKey
            : null;
        const /** @type {?} */ indexDbKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.key
            : null;
        const /** @type {?} */ data = arrayKey ? requestData[arrayKey] : requestData;
        switch (action) {
            case 'CREATE':
            case 'UPDATE':
                return (action === 'CREATE'
                    ? this.indexDbService.post(indexDbSchema, data)
                    : this.indexDbService.put(indexDbSchema, data)).pipe(map(() => requestData), catchError(e => {
                    console.warn('Could not save data into index DB, ERROR: ' + e);
                    return of(requestData);
                }));
            case 'DELETE':
                return this.indexDbService.delete(indexDbSchema, indexDbKey).pipe(map(() => requestData), catchError(e => {
                    console.warn('Could not delete data from index DB, ERROR: ' + e);
                    return of(requestData);
                }));
            default:
                console.log('No action has been specified');
                return of(requestData);
        }
    }
    /**
     * @param {?} httpConfig
     * @return {?}
     */
    _fetchFromIndexDb(httpConfig) {
        const /** @type {?} */ indexDbSchema = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.schema
            : null;
        const /** @type {?} */ indexDbKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.key
            : null;
        return indexDbKey
            ? this.indexDbService.findOne(indexDbSchema, indexDbKey)
            : this.indexDbService.findAll(indexDbSchema).pipe(map(response => {
                const /** @type {?} */ arrayKey = httpConfig.indexDbConfig
                    ? httpConfig.indexDbConfig.arrayKey
                    : null;
                return response.length > 0
                    ? arrayKey
                        ? { [arrayKey]: response }
                        : response
                    : null;
            }));
    }
}
NgxDhis2HttpClientService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
NgxDhis2HttpClientService.ctorParameters = () => [
    { type: HttpClient },
    { type: ManifestService },
    { type: SystemInfoService },
    { type: IndexDbService }
];
/** @nocollapse */ NgxDhis2HttpClientService.ngInjectableDef = defineInjectable({ factory: function NgxDhis2HttpClientService_Factory() { return new NgxDhis2HttpClientService(inject(HttpClient), inject(ManifestService), inject(SystemInfoService), inject(IndexDbService)); }, token: NgxDhis2HttpClientService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { ManifestService, SystemInfoService, NgxDhis2HttpClientService, IndexDbService };
