// @iapps/ngx-dhis2-menu/esm2015/lib/services/system-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { switchMap, tap } from "rxjs/operators";
import { BehaviorSubject, of, timer } from "rxjs";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class SystemStateService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this._loggingStatus$ = new BehaviorSubject(true);
        this._timeInterval = 30000;
        this._waitingTime = 0;
    }
    /**
     * @return {?}
     */
    checkOnlineStatus() {
        return timer(this._waitingTime, this._timeInterval).pipe(switchMap(() => of(navigator.onLine)), tap(isOnline => {
            if (isOnline) {
                // Set time interval to larger value if was lowered
                this._timeInterval = 30000;
                this._waitingTime = 500;
                this._checkLoginStatus(isOnline);
            }
            else {
                // Deduce time for the timer
                this._timeInterval = 500;
                this._waitingTime = 500;
            }
        }));
    }
    /**
     * @param {?} isOnline
     * @return {?}
     */
    _checkLoginStatus(isOnline) {
        this.pingServer().subscribe((pingResult) => {
            if (pingResult) {
                this._loggingStatus$.next(pingResult.loggedIn);
            }
        }, error => {
            if (isOnline) {
                this._loggingStatus$.next(false);
            }
        });
    }
    /**
     * @return {?}
     */
    getLoginStatus() {
        return this._loggingStatus$.asObservable();
    }
    /**
     * @return {?}
     */
    pingServer() {
        return this.httpClient.get("../../../dhis-web-commons-stream/ping.action");
    }
}
SystemStateService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
SystemStateService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ SystemStateService.ngInjectableDef = i0.defineInjectable({ factory: function SystemStateService_Factory() { return new SystemStateService(i0.inject(i1.HttpClient)); }, token: SystemStateService, providedIn: "root" });
function SystemStateService_tsickle_Closure_declarations() {
    /** @type {?} */
    SystemStateService.prototype._loggingStatus$;
    /** @type {?} */
    SystemStateService.prototype._timeInterval;
    /** @type {?} */
    SystemStateService.prototype._waitingTime;
    /** @type {?} */
    SystemStateService.prototype.httpClient;
}
