// @iapps/ngx-dhis2-http-client/esm5/lib/services/ngx-dhis2-http-client.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ManifestService } from './manifest.service';
import { SystemInfoService } from './system-info.service';
import { throwError, of } from 'rxjs';
import { catchError, map, mergeMap, switchMap } from 'rxjs/internal/operators';
import { HTTP_CONFIG } from '../constants/http-config.constant';
import { IndexDbService } from './index-db.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./manifest.service";
import * as i3 from "./system-info.service";
import * as i4 from "./index-db.service";
var NgxDhis2HttpClientService = /** @class */ (function () {
    function NgxDhis2HttpClientService(httpClient, manifestService, systemInfoService, indexDbService) {
        this.httpClient = httpClient;
        this.manifestService = manifestService;
        this.systemInfoService = systemInfoService;
        this.indexDbService = indexDbService;
    }
    /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype.get = /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    function (url, httpConfig) {
        var _this = this;
        var /** @type {?} */ newHttpConfig = tslib_1.__assign({}, HTTP_CONFIG, httpConfig);
        // Make a call directly from url if is external one
        if (newHttpConfig.isExternalLink) {
            return this.httpClient.get(url);
        }
        // load from index db
        if (newHttpConfig.useIndexDb) {
            return this._fetchFromIndexDb(newHttpConfig).pipe(mergeMap(function (indexDbResponse) {
                return indexDbResponse
                    ? of(indexDbResponse)
                    : _this._get(url, newHttpConfig).pipe(mergeMap(function (response) {
                        return _this._updateIndexDb(newHttpConfig, response);
                    }));
            }));
        }
        return this._get(url, newHttpConfig);
    };
    /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype.post = /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    function (url, data, httpConfig) {
        var _this = this;
        var /** @type {?} */ newHttpConfig = tslib_1.__assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap(function (rootUrl) {
            return _this.httpClient
                .post(rootUrl + url, data)
                .pipe(mergeMap(function (response) {
                var /** @type {?} */ dataId = response.uid;
                return newHttpConfig.useIndexDb
                    ? dataId
                        ? _this._updateIndexDb(newHttpConfig, tslib_1.__assign({}, data, { id: dataId }), 'CREATE').pipe(map(function () { return response; }))
                        : of(response)
                    : of(response);
            }))
                .pipe(catchError(_this._handleError));
        }), catchError(this._handleError));
    };
    /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype.put = /**
     * @param {?} url
     * @param {?} data
     * @param {?=} httpConfig
     * @return {?}
     */
    function (url, data, httpConfig) {
        var _this = this;
        var /** @type {?} */ newHttpConfig = tslib_1.__assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap(function (rootUrl) {
            return _this.httpClient.put(rootUrl + url, data).pipe(mergeMap(function (response) {
                return newHttpConfig.useIndexDb
                    ? _this._updateIndexDb(newHttpConfig, data, 'UPDATE').pipe(map(function () { return response; }))
                    : of(response);
            }), catchError(_this._handleError));
        }), catchError(this._handleError));
    };
    /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype.delete = /**
     * @param {?} url
     * @param {?=} httpConfig
     * @return {?}
     */
    function (url, httpConfig) {
        var _this = this;
        var /** @type {?} */ newHttpConfig = tslib_1.__assign({}, HTTP_CONFIG, httpConfig);
        return this._getSanitizedRootUrl(newHttpConfig).pipe(mergeMap(function (rootUrl) {
            return _this.httpClient
                .delete(rootUrl + url)
                .pipe(mergeMap(function (response) {
                return newHttpConfig.useIndexDb
                    ? _this._updateIndexDb(newHttpConfig, null, 'DELETE').pipe(map(function () { return response; }))
                    : of(response);
            }))
                .pipe(catchError(_this._handleError));
        }), catchError(this._handleError));
    };
    /**
     * @param {?} url
     * @param {?} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._get = /**
     * @param {?} url
     * @param {?} httpConfig
     * @return {?}
     */
    function (url, httpConfig) {
        var _this = this;
        return this._getSanitizedRootUrl(httpConfig).pipe(mergeMap(function (rootUrl) {
            return _this.httpClient.get(rootUrl + url).pipe(catchError(_this._handleError));
        }), catchError(this._handleError));
    };
    /**
     * @param {?} err
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._handleError = /**
     * @param {?} err
     * @return {?}
     */
    function (err) {
        var /** @type {?} */ error = null;
        if (err.error instanceof ErrorEvent) {
            // A client-side or network error occurred. Handle it accordingly.
            error = {
                message: err.error,
                status: err.status,
                statusText: err.statusText
            };
        }
        else {
            // The backend returned an unsuccessful response code.
            // The response body may contain clues as to what went wrong,
            error = {
                message: err.error instanceof Object
                    ? err.error.message
                    : err.error || err.message,
                status: err.status,
                statusText: err.statusText
            };
        }
        return throwError(error);
    };
    /**
     * @param {?} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._getSanitizedRootUrl = /**
     * @param {?} httpConfig
     * @return {?}
     */
    function (httpConfig) {
        return httpConfig.useRootUrl
            ? this.manifestService.getRootUrl()
            : this._getApiRootUrl(httpConfig.includeVersionNumber, httpConfig.preferPreviousApiVersion);
    };
    /**
     * @param {?=} includeVersionNumber
     * @param {?=} preferPreviousVersion
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._getApiRootUrl = /**
     * @param {?=} includeVersionNumber
     * @param {?=} preferPreviousVersion
     * @return {?}
     */
    function (includeVersionNumber, preferPreviousVersion) {
        var _this = this;
        if (includeVersionNumber === void 0) { includeVersionNumber = false; }
        if (preferPreviousVersion === void 0) { preferPreviousVersion = false; }
        var /** @type {?} */ rootUrlPromise = this.manifestService.getRootUrl().pipe(switchMap(function (rootUrl) {
            return _this.systemInfoService.getSystemVersion().pipe(map(function (version) {
                return {
                    rootUrl: rootUrl,
                    version: version - 1 <= 25 ? version + 1 : version
                };
            }));
        }));
        return rootUrlPromise.pipe(map(function (urlInfo) {
            return urlInfo.rootUrl + "api/" + (includeVersionNumber && !preferPreviousVersion
                ? urlInfo.version + '/'
                : preferPreviousVersion
                    ? urlInfo.version
                        ? urlInfo.version - 1 + '/'
                        : ''
                    : '');
        }));
    };
    /**
     * @param {?} httpConfig
     * @param {?} requestData
     * @param {?=} action
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._updateIndexDb = /**
     * @param {?} httpConfig
     * @param {?} requestData
     * @param {?=} action
     * @return {?}
     */
    function (httpConfig, requestData, action) {
        if (action === void 0) { action = 'CREATE'; }
        var /** @type {?} */ indexDbSchema = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.schema
            : null;
        var /** @type {?} */ arrayKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.arrayKey
            : null;
        var /** @type {?} */ indexDbKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.key
            : null;
        var /** @type {?} */ data = arrayKey ? requestData[arrayKey] : requestData;
        switch (action) {
            case 'CREATE':
            case 'UPDATE':
                return (action === 'CREATE'
                    ? this.indexDbService.post(indexDbSchema, data)
                    : this.indexDbService.put(indexDbSchema, data)).pipe(map(function () { return requestData; }), catchError(function (e) {
                    console.warn('Could not save data into index DB, ERROR: ' + e);
                    return of(requestData);
                }));
            case 'DELETE':
                return this.indexDbService.delete(indexDbSchema, indexDbKey).pipe(map(function () { return requestData; }), catchError(function (e) {
                    console.warn('Could not delete data from index DB, ERROR: ' + e);
                    return of(requestData);
                }));
            default:
                console.log('No action has been specified');
                return of(requestData);
        }
    };
    /**
     * @param {?} httpConfig
     * @return {?}
     */
    NgxDhis2HttpClientService.prototype._fetchFromIndexDb = /**
     * @param {?} httpConfig
     * @return {?}
     */
    function (httpConfig) {
        var /** @type {?} */ indexDbSchema = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.schema
            : null;
        var /** @type {?} */ indexDbKey = httpConfig.indexDbConfig
            ? httpConfig.indexDbConfig.key
            : null;
        return indexDbKey
            ? this.indexDbService.findOne(indexDbSchema, indexDbKey)
            : this.indexDbService.findAll(indexDbSchema).pipe(map(function (response) {
                var /** @type {?} */ arrayKey = httpConfig.indexDbConfig
                    ? httpConfig.indexDbConfig.arrayKey
                    : null;
                return response.length > 0
                    ? arrayKey
                        ? (_a = {}, _a[arrayKey] = response, _a) : response
                    : null;
                var _a;
            }));
    };
    NgxDhis2HttpClientService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] },
    ];
    /** @nocollapse */
    NgxDhis2HttpClientService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: ManifestService },
        { type: SystemInfoService },
        { type: IndexDbService }
    ]; };
    /** @nocollapse */ NgxDhis2HttpClientService.ngInjectableDef = i0.defineInjectable({ factory: function NgxDhis2HttpClientService_Factory() { return new NgxDhis2HttpClientService(i0.inject(i1.HttpClient), i0.inject(i2.ManifestService), i0.inject(i3.SystemInfoService), i0.inject(i4.IndexDbService)); }, token: NgxDhis2HttpClientService, providedIn: "root" });
    return NgxDhis2HttpClientService;
}());
export { NgxDhis2HttpClientService };
function NgxDhis2HttpClientService_tsickle_Closure_declarations() {
    /** @type {?} */
    NgxDhis2HttpClientService.prototype.httpClient;
    /** @type {?} */
    NgxDhis2HttpClientService.prototype.manifestService;
    /** @type {?} */
    NgxDhis2HttpClientService.prototype.systemInfoService;
    /** @type {?} */
    NgxDhis2HttpClientService.prototype.indexDbService;
}
